### ISSUE TITLE
Micosoft's OAuth token endpoint expects form-data, not JSON

### ISSUE THREAD

**Issue by kbni** [cronicle-edge/cronicle-edge#165]
Microsoft's Oauth2 implementation requires the [request for a token](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow#request-an-access-token-with-a-client_secret) to be sent as form data.

I tried to make user.js send form data but was pretty unsuccessful figuring out how to pass the correct options through to pixl-request:

```
			var tokenResponse = null;
			const tokenRequestPayload = Tools.mergeHashes({
				client_id: oauth.client_id,
				client_secret: oauth.client_secret,
				code: code,
				grant_type: 'authorization_code',
				redirect_uri: oauth.redirect_uri
			}, oauth.params);

			if(oauth.token_url.startsWith('https://login.microsoftonline.com/')) {
				self.logDebug(3, 'Using form data for Microsoft oauth')
				let payloadFormBody = new URLSearchParams(tokenRequestPayload).toString();
				console.log("payloadFormBody:", payloadFormBody);
				tokenResponse = await self.postJsonAsync(oauth.token_url, {
					data: Buffer.from(payloadFormBody),
					headers: { 'content-type': 'application/x-www-form-urlencoded' }
				});
			}
```

Is there any chance you can tell me what I've done wrong above? Seems like postJsonAsync should just pass options directly through to pixl-request, but it doesn't seem to.

**Comment by kbni** [comment]
Update: seems something else was going wrong (using incorrect user_info URL), ended up getting this working with a few [modifications to user.js](https://github.com/kbni/cronicle-edge/commit/2b02c00f221aa3fa51549e02113a9477a4ded459) and the following configuration values:

```
  "oauth": {
    "enabled": true,
    "insecure": false,
    "client_id": "<< YOUR APPLICATION ID >>",
    "client_secret": "<< REDACTED >>",
    "redirect_uri": "https://cr.dorks.com.au/api/user/callback",
    "authorize_url": "https://login.microsoftonline.com/<<YOUR TENANT ID>>/oauth2/v2.0/authorize",
    "token_url": "https://login.microsoftonline.com/<<YOUR TENANT ID>>/oauth2/v2.0/token",
    "user_url": "https://graph.microsoft.com/v1.0/me",
    "user_attribute": "email",
    "scope": "email openid profile",
    "user_attribte": "userPrincipalName"
  },
```

Very cool project, thank you! :)

**Comment by mikeTWC1984** [comment]
OK, good.

So I tried it myself, it seemed to work without any modification. I know the token API is the place where providers might have some differences, so I have extra config to inject query parameters and headers to that request (**params and headers**) to that request. E.g. you can add content-type header like this:

```json
	"oauth": {
		"enabled": true,
		"insecure": false,
		"client_id": "XXXXXXXXXX",
		"client_secret": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
		"redirect_uri": "http://localhost:3012/api/user/callback",
		"authorize_url": "http://localhost:9000/application/o/authorize/",
		"token_url": "http://localhost:9000/application/o/token/",
		"user_url": "http://localhost:9000/application/o/userinfo/",
		"user_attribute": "preferred_username",
		"avatar_attribute": "avatar_url",
		"scope": "email openid profile",
		"params": {
			"access_type": "offline",
			"prompt": "consent"
		},
		"headers": {
			"accept": "application/json",
                        "content-type": "application/x-www-form-urlencoded"
		}
	}
```

**Comment by fdquinones** [comment]
> OK, good.
> 
> So I tried it myself, it seemed to work without any modification. I know the token API is the place where providers might have some differences, so I have extra config to inject query parameters and headers to that request (**params and headers**) to that request. E.g. you can add content-type header like this:
> 
> 	"oauth": {
> 		"enabled": true,
> 		"insecure": false,
> 		"client_id": "XXXXXXXXXX",
> 		"client_secret": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
> 		"redirect_uri": "http://localhost:3012/api/user/callback",
> 		"authorize_url": "http://localhost:9000/application/o/authorize/",
> 		"token_url": "http://localhost:9000/application/o/token/",
> 		"user_url": "http://localhost:9000/application/o/userinfo/",
> 		"user_attribute": "preferred_username",
> 		"avatar_attribute": "avatar_url",
> 		"scope": "email openid profile",
> 		"params": {
> 			"access_type": "offline",
> 			"prompt": "consent"
> 		},
> 		"headers": {
> 			"accept": "application/json",
>                         "content-type": "application/x-www-form-urlencoded"
> 		}
> 	}

@mikeTWC1984, I agree with you. For me, the integration with Microsoft auth was successful without needing to change any code.
